samza-api/src/main/java/org/apache/samza/context/Context.java

```diff
@@ -0,0 +1,42 @@
+package org.apache.samza.context;
+
+/**
+ * Container object for all context provided to instantiate an application at runtime.
+ */
+public interface Context {
+  /**
+   * @return framework-provided context for the overall job that is being run
+   */
+  JobContext getJobContext();
+
+  /**
+   * @return framework-provided context for the container that this is in
+   */
+  ContainerContext getContainerContext();
+
+  /**
+   * @return framework-provided context for the task that that this is in
+   */
+  TaskContext getTaskContext();
+
+  /**
+   * Returns the application-defined container context object specified by the
+   * {@link ApplicationDefinedContainerContextFactory}. This is shared across all tasks in the container, but not across
+   * containers.
+   * @param clazz concrete type of the {@link ApplicationDefinedContainerContext} to return
+   * @return application-defined container context
+   * @throws IllegalStateException if no factory was provided
+   * @throws ClassCastException if the type of the concrete context does not match clazz
+   */
+  <T extends ApplicationDefinedContainerContext> T getApplicationDefinedContainerContext(Class<T> clazz);
```

Comment:
Is the class parameter just for knowing the concrete type? If so, I'm not sure this is better then requiring them to cast it themselves.

Reply:
Slightly prefer casting for simplicity and consistency with getTable and getStore APIs.